### PR TITLE
Fixed #26192 -- Fixed crash of ordering by constants on PostgreSQL.

### PR DESCRIPTION
**Pull request by felixxm**
Ordering by constants is reasonable in some cases (e.g. in compound queries), but in the same time it can be problematic on PostgreSQL without specified output type (see ticket [26192](https://code.djangoproject.com/ticket/26192)).

IMO we should raise `FieldError` when `output_field` cannot be resolved.

@charettes What do you think?

### CODE REVIEW

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by cansarigol** [comment]
I remembered @charettes [pr](https://github.com/django/django/pull/11359) about this.

**Comment by charettes** [comment]
@felixxm I'm not sure this is the right approach.

Somewhere between when the ticket was initially created (1.8 era) and now the `annotate(constant=Value('foo')).order_by('constant')` expression changed from being compiled to

```sql
SELECT ..., 'foo' AS constant ORDER BY 'foo'
```

to 

```sql
SELECT ..., 'foo' AS constant ORDER BY constant
```

The main difference now is that `order_by` reuses existing aliases which explains why your `test_order_by_constant_value` test passes. The original issue can still be reproduced by passing an expression directly to `order_by` though even if `output_field` is specified:

```python
def test_order_by_constant_value(self):
    qs = Article.objects.order_by(
        constant=Value('1', output_field=IntegerField()),
    )
    list(qs)
``` 

The reason for this is that `output_field` resolution doesn't result in a cast (e.g. `::text`) which is what seems to be tripping PostgreSQL here. Also the `test_order_by_constant_value_without_output_field` will to fail if https://github.com/django/django/pull/11359 is merged as @cansarigol pointed out.

Here's my take on how this could be resolved:

1. Make `order_by` simply ignore `Value` expressions. They are constants anyway so there's no point in ordering by them.
2. Require that `Value` expressions passed to `order_by` be resolvable like you did here and wrap them in ` Cast(val, val.output_field)`.

**Comment by felixxm** [comment]
@charettes Thanks for your analysis :bowing_man: I added `Cast()` expressions and extra tests.

I agree that original issue can be reproduced by passing an expression directly to `order_by` but only when `output_field` is not an `IntegerField`, because integers are interpreted as column numbers (what is misleading).

I also know `test_order_by_constant_value_without_output_field` will not raise exceptions after #11359 but that's OK (IMO) we can move these use cases into `test_order_by_constant_value`.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/ordering/tests.py [hunk 1]
```diff
@@ -402,6 +405,37 @@ def test_order_by_f_expression_duplicates(self):
...
         )
 
+    def test_order_by_constant_value(self):
+        # Order by annotated constant from selected columns.
+        qs = Article.objects.annotate(
+            constant=Value('1', output_field=CharField()),
+        ).order_by('constant', '-headline')
+        self.assertSequenceEqual(qs, [self.a4, self.a3, self.a2, self.a1])
+        # Order by annotated constant which is out of selected columns.
+        self.assertQuerysetEqual(
```
Comment (charettes) [inline tests/ordering/tests.py]: Could use `assertSequenceEqual` to avoid the `itemgetter`

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -299,10 +303,13 @@ def get_order_by(self):
                     True))
                 continue
             if col in self.query.annotations:
-                # References to an expression which is masked out of the SELECT clause
-                order_by.append((
-                    OrderBy(self.query.annotations[col], descending=descending),
-                    False))
+                # References to an expression which is masked out of the SELECT
+                # clause.
+                col = self.query.annotations[col]
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: I think `annotation` or `expr` might be a more appropriate variable name.

**Comment by felixxm** [comment]
@charettes Thanks!

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
buildbot

**Pull request merged**